Sources/NIOCore/ByteBuffer-quicBinaryEncodingStrategy.swift

```diff
@@ -0,0 +1,144 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+extension ByteBuffer {
+    /// A ``NIOBinaryIntegerEncodingStrategy`` which encodes bytes as defined in RFC 9000 § 16
+    public struct QUICBinaryEncodingStrategy: NIOBinaryIntegerEncodingStrategy {
+        /// An estimate of the bytes required to write integers using this strategy
+        public var requiredBytesHint: Int
+
+        /// - Parameter requiredBytesHint: An estimate of the bytes required to write integers using this strategy
+        @inlinable
+        public init(requiredBytesHint: Int) {
+            precondition(
+                requiredBytesHint == 0
+                    || requiredBytesHint == 1
+                    || requiredBytesHint == 2
+                    || requiredBytesHint == 4
+                    || requiredBytesHint == 8
+            )
+            self.requiredBytesHint = requiredBytesHint
+        }
+
+        @inlinable
+        public func readInteger<IntegerType: FixedWidthInteger>(
+            as: IntegerType.Type,
+            from buffer: inout ByteBuffer
+        ) -> IntegerType? {
+            guard let firstByte = buffer.getInteger(at: buffer.readerIndex, as: UInt8.self) else {
+                return nil
+            }
+
+            // Look at the first two bits to work out the length, then read that, mask off the top two bits, and
+            // extend to integer.
+            switch firstByte & 0xC0 {
+            case 0x00:
+                // Easy case.
+                buffer.moveReaderIndex(forwardBy: 1)
+                return IntegerType(firstByte & ~0xC0)
+            case 0x40:
+                // Length is two bytes long, read the next one.
+                return buffer.readInteger(as: UInt16.self).map { IntegerType($0 & ~(0xC0 << 8)) }
+            case 0x80:
+                // Length is 4 bytes long.
+                return buffer.readInteger(as: UInt32.self).map { IntegerType($0 & ~(0xC0 << 24)) }
+            case 0xC0:
+                // Length is 8 bytes long.
+                return buffer.readInteger(as: UInt64.self).map { IntegerType($0 & ~(0xC0 << 56)) }
+            default:
+                fatalError("Unreachable")
+            }
+        }
+
+        private enum NumBytes {
+            case one, two, four, eight
+        }
+
+        @usableFromInline
+        func bytesNeededForInteger<IntegerType: FixedWidthInteger>(_ integer: IntegerType) -> Int {
```

Comment:
I think exposing this logic publicly would be useful. Consider a case where I want to write multiple variable length integers to a buffer - how can I appropriately size the buffer's capacity ahead of time? I can assume they'll all need 8 bytes, but then I'm going to over allocate. 
```
var buffer = allocator.buffer(capacity: enoughRoomForTwoVarints)
buffer.writeInteger(myInt, strategy: .quic)
buffer.writeInteger(myOtherInt, strategy: .quic)
```